Require types to be conformant for eta-expansion of SAM under -Xsource:3

eed3si9n opened the pull request:
Ref https://github.com/scala/bug/issues/12006
Ref https://github.com/scala/scala/pull/7660

```scala
class C(s: java.io.InputStream); class D(u: java.net.URL) extends C(u.openStream)
```

raised in https://github.com/scala/bug/issues/12006 a weird piece of code under -Xsource:3 since we can interpret it in two ways.

1. `InputStream` is a SAM of `def read(): Int`, and `C(u.openStream)` requires an eta-expansion from openStream method to `() => Int`.
2. `u.openStream` needs auto application `()`.

Under -Xsource:3 Scala 2.13.2 takes the first interpretation since https://github.com/scala/scala/pull/7660 and then errors with surprising:

```
       error: type mismatch;
        found   : java.io.InputStream
        required: Int
```

This PR adds an additional check to see if the types match up between `def read(): Int` and `URL#openStream` before saying yes to eta-expansion.

som-snytt commented:
That was the fastest merge conflict ever.

Nicely done!

Why does `a function-ish type of arity 0 including SAM` sound like a Monty Python sketch?

eed3si9n commented:
> That was the fastest merge conflict ever.

My 2.13.x branch was very out of date. Rebased to latest now.

Jasper-M commented:
I think you should differentiate between `=> InputStream` and `() => InputStream`.

dwijnand reviewed (commented):
Dotty (under all flags: `-source 3.0-migration`, `-source 3.0` (aka no flag), `-source 3.1-migration`, and `-source 3.1`) emits a warning with the message: "method `<foo>` must be called with () argument" (with a nice explanation with `-explain`).

So it would be slightly better if those AcciSamZero/SamZero didn't revert back to just "type mismatch" messaging, but let me know if you'd rather land this as is.

eed3si9n commented:
@Jasper-M 

> I think you should differentiate between `=> InputStream` and `() => InputStream`.

Good point. It's a good indication that something is wrong when removing eta expansion fixes a by-name parameter issue. I guess at bytecode level `class C(s: => java.io.InputStream)` is represented using `Function0`?

```java
scala> :javap C
....
{
  public C(scala.Function0<java.io.InputStream>);
    descriptor: (Lscala/Function0;)V
    flags: ACC_PUBLIC
    Code:
      stack=1, locals=2, args_size=2
         0: aload_0
         1: invokespecial #15                 // Method java/lang/Object."<init>":()V
         4: return
      LineNumberTable:
        line 1: 0
      LocalVariableTable:
        Start  Length  Slot  Name   Signature
            0       5     0  this   LC;
            0       5     1     s   Lscala/Function0;
    Signature: #11                          // (Lscala/Function0<Ljava/io/InputStream;>;)V
    MethodParameters:
      Name                           Flags
      s                              final
}
```

@dwijnand Let me see if we can save the baby from the bath water.

eed3si9n commented:
Actually by-name (I think) is a red herring. Since I can make this fail without it:

```scala
scala> class C(s: java.io.InputStream); class D(u: java.net.URL) extends C(u.openStream)
                                                                             ^
       error: type mismatch;
        found   : java.io.InputStream
        required: Int
```

From adding a few printlns:

```scala
scala> class C(s: java.io.InputStream); class D(u: java.net.URL) extends C(u.openStream)
ptUnderlying java.io.InputStream
sam method read
                                                                             ^
       error: type mismatch;
        found   : java.io.InputStream
        required: Int
```

Typer thinks we need `java.io.InputStream`, and tries to convert `URL#openStream` into `InputStream#read`, which might explain where `Int` is coming from.

eed3si9n committed:
Require types to be conformant for eta-expansion of SAM under -Xsource:3

Ref scala bug 12006
Ref scala scala 7660

```
class C(s: java.io.InputStream); class D(u: java.net.URL) extends C(u.openStream)
```

raised in 12006 a weird piece of code under -Xsource:3 since we can interpret it in two ways.

1. `InputStream` is a SAM of `def read(): Int`, and `C(u.openStream)` requires an eta-expansion from openStream method to `() => Int`.
2. `u.openStream` needs auto application `()`.

Under -Xsource:3 Scala 2.13.2 takes the first interpretation since 7660 and then errors with surprising:

```
   error: type mismatch;
    found   : java.io.InputStream
    required: Int
```

This adds an additional check to see if the types weakly match up between `def read(): Int` and `URL#openStream` before saying yes to eta-expansion.

eed3si9n commented:
In a5f302d I did:

```diff
          def expectingSamOfArity = {
            val sam = samOf(ptUnderlying)
-           sam.exists && sam.info.params.lengthCompare(arity) == 0
+           sam.exists && (sam.info.params.lengthCompare(arity) == 0) && (mt weak_<:< sam.info)
          }
```

This results in the following errors when starr is changed https://travis-ci.org/github/scala/scala/jobs/693918166

```scala
[error] /home/travis/build/scala/scala/src/reflect/scala/reflect/internal/Types.scala:4466:80: missing argument list for method infoTypeDepth in trait Types
[error] Unapplied methods are only converted to functions when a function type is expected.
[error] You can make this conversion explicit by writing `infoTypeDepth _` or `infoTypeDepth(_)` instead of `infoTypeDepth`.
[error]   private def symTypeDepth(syms: List[Symbol]): Depth  = Depth.maximumBy(syms)(infoTypeDepth)
[error]                                                                                ^
[error] /home/travis/build/scala/scala/src/reflect/scala/reflect/internal/Types.scala:5291:26: missing argument list for method typeDepth in trait Types
[error] Unapplied methods are only converted to functions when a function type is expected.
[error] You can make this conversion explicit by writing `typeDepth _` or `typeDepth(_)` instead of `typeDepth`.
[error]     Depth.maximumBy(tps)(typeDepth)
[error]                          ^
```

eed3si9n commented on src/compiler/scala/tools/nsc/typechecker/Typers.scala:
```diff
@@ -940,7 +940,7 @@ trait Typers extends Adaptations with Tags with TypersTracking with PatternTyper
           def expectingSamOfArity = {
             val sam = samOf(ptUnderlying)
-            sam.exists && sam.info.params.lengthCompare(arity) == 0
+            sam.exists && (sam.info.params.lengthCompare(arity) == 0) && (mt weak_<:< sam.info)
```
I tried `(mt matches sam.info)` here and that doesn't work either

```diff
+t7187-3.scala:39: error: type mismatch;
+ found   : java.io.InputStream
+ required: Int
+  class D1(u: java.net.URL) extends C1(u.openStream) // ok
+                                         ^
+t7187-3.scala:42: error: type mismatch;
+ found   : java.io.InputStream
+ required: Int
+  class D2(u: java.net.URL) extends C2(u.openStream) // ok
+                                         ^
```

eed3si9n commented:
Got green on Travis CI (https://travis-ci.org/github/scala/scala/builds/694444308), so I'm squashing the commits.

lrytz commented on src/compiler/scala/tools/nsc/typechecker/Typers.scala:
```diff
@@ -938,9 +938,17 @@ trait Typers extends Adaptations with Tags with TypersTracking with PatternTyper
+            sam.exists &&
+            (if (typeCheck)
+              sam.info match {
+                case mt2: MethodType => isApplicableMethod(mt, mt2)
```
this should only be checked if arity is 0, no? so it would be enough to check return types.

eed3si9n replied:
Yes. I went back and forth on this PR. At some point I was trying to do the check with or without -Xsource:3. I can bring the code back in here to clarify that it's only doing return type check.

lrytz reviewed (changes requested):
I'm skeptical about the approach. For example, the following is currently allowed (under -Xsource:3), but will no longer work

```
@FunctionalInterface trait Sam { def f(): Int }

object U {
  import scala.language.implicitConversions
  implicit def s2i(s: Option[Int]): Int = s.getOrElse(0)
  def s(): Option[Int] = ???
  val x: Sam = s
}
```

I'm also very confused about dotty. The following compiles with scalac -Xsource:3, and it should work on dotty, right? But it doesn't seem to.

```
@FunctionalInterface trait Sam { def f(): Int }
object T {
  def f(): Int = ???
  val x: Sam = f
}
```

gives

```
-- [E100] Syntax Error: Test.scala:4:15 ----------------------------------------
4 |  val x: Sam = f
  |               ^
  |               method f must be called with () argument

longer explanation available when compiling with `-explain`
1 error found
```

Finally, for the example that we're actually trying to fix

```
trait Sam { def f(): Int }
object T {
  def sam(): Sam = ???
  val x: Sam = sam
}
```

the current situation in 2.13 (with -Xsource:3) would already be better if the warning

```
warning: Eta-expansion performed to meet expected type Sam, which is SAM-equivalent to () => Int,
even though trait Sam is not annotated with `@FunctionalInterface`;
to suppress warning, add the annotation or write out the equivalent function literal.
```

was showing up. Unfortunately the warning is hidden because the `found: Sam, required: Int` error has the same position (it shows under `-Vdebug`), see https://github.com/scala/scala/blob/c5619f2c5d7e7ec80724d29e56a97bcd640c546c/src/compiler/scala/tools/nsc/reporters/Reporter.scala#L108-L114.

eed3si9n commented:
@lrytz What do you suggest `-Xsource:3` do?  I would say that it would be least surprising if `-Xsource:3` simply dropped SAM eta-expansion support given it doesn't seem to be working anyway in Dotty.

lrytz commented:
I [asked on gitter](https://gitter.im/lampepfl/dotty?at=5ede3d5a2c49c45f5acdaf49) and @smarter suggests it might be an oversight. So I created https://github.com/lampepfl/dotty/issues/9135 to ask for a clarification.

lrytz commented:
Martin says it's by design, no eta expansion of nilary methods is only done when the expected type is `() => T`, not a SAM. So we have to adjust what scalac does under `-Xsource:3`.

som-snytt commented:
Thank-you all for pursuing this. I meant to allocate some time for it, but my tail is already losing fur from chasing it. Also, in lockdown, time has no meaning. If it's OK, I will not delete my comment that it sounds like the title of a Monty Python skit.

eed3si9n commented:
lol I should've kept my first commit that removed eta expansion of SAM.

<img width="774" alt="Require_types_to_be_conformant_for_eta-expansion_of_SAM_under_-Xsource_3_by_eed3si9n_·_Pull_Request__9027_·_scala_scala" src="https://user-images.githubusercontent.com/184683/84159379-014f7b80-aa3b-11ea-9cb1-21eadd4f1504.png">
